### DISCUSSION TITLE
markdown grammar broken

### DISCUSSION THREAD

**Discussion by lunasophia** [nvim-treesitter/nvim-treesitter#6618] [General]
### Describe the bug

When I enable `markdown`, highlighting fails in markdown files. It's really that simple. If I enable `markdown` or `markdown_inline`, highlighting breaks. As soon as I add the two grammars to the list of disabled parsers, highlighting works again.

### To Reproduce

Install nvim-treesitter, enable markdown, open markdown file, all highlighting fails.

### Expected behavior

Highlighting works

### Output of `:checkhealth nvim-treesitter`

```markdown
==============================================================================
nvim-treesitter: require("nvim-treesitter.health").check()

Installation ~
- OK `tree-sitter` found 0.22.5 (parser generator, only needed for :TSInstallFromGrammar)
- WARNING `node` executable not found (only needed for :TSInstallFromGrammar, not required for :TSInstall)
- OK `git` executable found.
- OK `cc` executable found. Selected from { vim.NIL, "cc", "gcc", "clang", "cl", "zig" }
  Version: cc (GCC) 14.1.1 20240507 (Red Hat 14.1.1-1)
- OK Neovim was compiled with tree-sitter runtime ABI version 14 (required >=13). Parsers must be compatible with runtime ABI.

OS Info:
{
  machine = "aarch64",
  release = "6.8.9-402.asahi.fc40.aarch64+16k",
  sysname = "Linux",
  version = "#1 SMP PREEMPT_DYNAMIC Wed May  8 18:05:42 UTC 2024"
} ~

Parser/Features         H L F I J
  - c                   ✓ ✓ ✓ ✓ ✓
  - gitcommit           ✓ . . . ✓
  - lua                 ✓ ✓ ✓ ✓ ✓
  - markdown            ✓ . ✓ ✓ ✓
  - python              ✓ ✓ ✓ ✓ ✓
  - query               ✓ ✓ ✓ ✓ ✓
  - toml                ✓ ✓ ✓ ✓ ✓
  - vim                 ✓ ✓ ✓ . ✓
  - vimdoc              ✓ . . . ✓

  Legend: H[ighlight], L[ocals], F[olds], I[ndents], In[j]ections
         +) multiple parsers found, only one will be used
         x) errors found in the query, try to run :TSUpdate {lang} ~
```

### Output of `nvim --version`

```text
NVIM v0.9.5
Build type: RelWithDebInfo
LuaJIT 2.1.1707061634
Compilation: /usr/bin/gcc -O2 -g -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wvla -Wdouble-promotion -Wmissing-noreturn -Wmissing-format-attribute -Wmissing-prototypes -fno-common -Wno-unused-result -Wimplicit-fallthrough -fdiagnostics-color=auto -fstack-protector-strong -DUNIT_TESTING -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DUSING_UV_SHARED=1 -I/usr/include/luajit-2.1 -I/usr/include -I/usr/include/luv -I/builddir/build/BUILD/neovim-0.9.5/redhat-linux-build/src/nvim/auto -I/builddir/build/BUILD/neovim-0.9.5/redhat-linux-build/include -I/builddir/build/BUILD/neovim-0.9.5/redhat-linux-build/cmake.config -I/builddir/build/BUILD/neovim-0.9.5/src -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include

system vimrc file: "$VIM/sysinit.vim"
fall-back for $VIM: "/usr/share/nvim"

Run :checkhealth for more info
```

### Additional context

Other languages, like `python`, do provide proper highlighting with nvim-treesitter. It's just markdown that's broken.

I thought installing nodejs fixed things, but realized I made a mistake with my config. Installing nodejs doesn't fix things, the grammar is just broken on my machine.

**Comment by clason** [comment]
No, that's not what's happening here. You also need to install the `markdown_inline` parser; both are needed for markdown parsing.

**Comment by lunasophia** [comment]
Yes, it is what's happening here. I have both.

**Comment by clason** [comment]
No it's not. And please provide minimal reproducing steps. The markdown parser works perfectly fine for everyone else.

**Comment by lunasophia** [comment]
How can I provide more minimal reproducing steps than "install it and it breaks"? Furthermore, I have installed `markdown_inline` and it's still broken. The flippant "no you're wrong" response is really discouraging.

![image](https://github.com/nvim-treesitter/nvim-treesitter/assets/104850249/8a4ed5b2-fd3e-43bb-881f-ec7459cb223c)

**Comment by topocount** [comment]
I've had this issue for months as well and have had to fall back to neovim's builtin markdown parsing

**Reply by lunasophia** [reply]
I don't think anything's going to happen with this, sadly, since the developers clearly think they're infallible and no one could ever find a problem with their precious software. 🤷‍♀️

**Reply by tuurep** [reply]
Could it have anything to do with your colorschemes, and the change in these capture groups' names?

https://github.com/nvim-treesitter/nvim-treesitter/commit/1ae9b0e4558fe7868f8cda2db65239cfb14836d0

(trying to actually think what's wrong with you guys' setups instead of being childish :smile:)

**Comment by topocount** [comment]
Everyone here who's still having issues or using a workaround might want to check out how neovim is packaged by their package manager if they're using linux.

I definitely figured out this problem (for myself, at least). I used archlinux (btw) and it started shipping neovim with a handful of (formerly optional, now mandatory) treesitter parsers as dependencies, including markdown. I pinned my neovim install to 0.9, which doesn't require treesitter parsers to be installed at the root level as dependencies of neovim, tacitly uninstalled the root-level markdown parser I installed, utilized my user-level `:TSInstall markdown` and was good again, everything working as expected. 

To reiterate, this is most likely a packaging issue, or less charitably an issue with how Treesitter handles parsers installed at both the root and user level.

**Comment by JBongars** [comment]
Hi all,

Leaving this here in case people need it.

I have been facing the same issues on Arch (btw) and MacOS. Same error so this is not an isolated case. You can download my dotfiles for nvim and try on the latest version of nvim (0.12.4) and it fails.

https://github.com/JBongars/dotfiles-nvim/tree/aeea314193219efe1b9feed45e97833d49de59a8

There are two possible solutions that worked for me. 

Option 1: You can fallback to using vim's grep pattern in case tree-sitter fails on any version of treesitter (I think). Modify you treesitter config like so

```lua
return { -- Highlight, edit, and navigate code
  'nvim-treesitter/nvim-treesitter',
  build = ':TSUpdate',
  main = 'nvim-treesitter.configs', -- Sets main module to use for opts
  branch = 'master',
  -- [[ Configure Treesitter ]] See `:help nvim-treesitter`
  opts = {
    ensure_installed = {
      'bash',
      -- truncated --
      'markdown',
      'markdown_inline',
    },

    -- Autoinstall languages that are not installed
    auto_install = true,
    highlight = {
      enable = true,
      additional_vim_regex_highlighting = { 'ruby', 'markdown' },  -- < ===== add 'markdown' here [!!]
    },
    indent = { enable = true, disable = { 'ruby' } },
  },
}
```

Option 2: You can downgrade your nvim version back to version 0.11.7. I did this via bob so you can periodically hop back and forth in case there are new changes pushed to neovim.

```bash
# install bob-nvim 
# brew install bob
# sudo pacman -S extra/bob
# choco install bob -- I'm guessing?

bob install v0.11.7
bob use v0.11.7

# symlink so you can use bob use without running again...
ln -s ~/.local/share/bob/nvim-bin/nvim ~/.local/bin/nvim 

# alternatively, permanently install
sudo cp ~/.local/share/bob/nvim-bin/nvim /usr/local/bin
```

Hope this helps!

**Comment by gx089** [comment]
I think everyone should read the docs and use the new way of configuring Tree-sitter. Use the new way of configuring Tree-sitter:

There is no need to install markdown or markdown_inline separately, as they are bundled with Neovim itself.
A Markdown file that includes code blocks with other languages requires the corresponding languages/parsers to be installed and enabled as well.


```lua
-- Install nvim-ts... this part may be different in older versions of neovim without vim.pack
vim.pack.add({
    { src = "https://github.com/nvim-treesitter/nvim-treesitter" },
})

-- Install the languages
require('nvim-treesitter').install({
    'cpp', 'zig', 'perl', 'javascript', 'typescript', 'tsx', "html", "css", "angular"
})

-- Important!
-- Activate TS for those languages. Additional languages are not enabled automatically
vim.api.nvim_create_autocmd('FileType', {
    pattern = {
        'cpp', 'zig', 'perl',
        'javascript', 'typescript', 'javascriptreact', 'typescriptreact', 'tsx', "html", "css", "htmlangular"
    },
    callback = function() vim.treesitter.start() end
})


```

**Reply by JBongars** [reply]
Thank you for your code snippet. May I check what version of nvim you are using?

**Reply by gx089** [reply]
I’m using a dev version of 0.13, but this definitely works with your version, 0.12.4.
